### ISSUE TITLE
[Accepted with Revisions] SDL 0141 - Supporting simultaneous multiple transports

### ISSUE THREAD

**Issue by theresalech** [smartdevicelink/sdl_evolution#405]
Hello SDL community,

The review of the revised proposal "SDL 0141 - Supporting simultaneous multiple transports" begins now and runs through March 13, 2018. The original review of "Supporting simultaneous multiple transports" occurred February 13 through February 20, 2018. The proposal is available here:

https://github.com/smartdevicelink/sdl_evolution/blob/master/proposals/0141-multiple-transports.md

**Please be aware that this proposal is contingent upon others, as noted within the proposal file.**

Reviews are an important part of the SDL evolution process. All reviews should be sent to the associated Github issue at:

https://github.com/smartdevicelink/sdl_evolution/issues/405

What goes into a review?

The goal of the review process is to improve the proposal under review through constructive criticism and, eventually, determine the direction of SDL. When writing your review, here are some questions you might want to answer in your review:

* Is the problem being addressed significant enough to warrant a change to SDL?
* Does this proposal fit well with the feel and direction of SDL?
* If you have used competitors with a similar feature, how do you feel that this proposal compares to those?
* How much effort did you put into your review? A glance, a quick reading, or an in-depth study?
Please state explicitly whether you believe that the proposal should be accepted into SDL.

More information about the SDL evolution process is available at

https://github.com/smartdevicelink/sdl_evolution/blob/master/process.md

Thank you,
Theresa Lech

Program Manager - Livio
theresa@livio.io

**Comment by joeljfischer** [comment]
### Alternate Transport Flow

Why was the decision made to connect all services across all transports instead of picking and choosing which transport will connect to which transport?

For the iOS proxy I would much prefer the following flow:

1. iOS proxy connects over iAP
2. RPC start service response contains the IP address / port, and which service numbers should support which transport. The RPC transport would always use the IAP transport.
3. If the IAP transport disconnects, all other transports disconnect.

I feel that this would eliminate *a lot* of complexity and edge cases. It's not quite as powerful with the transport switching and such, but it's far less complex, and I don't feel that the extra power is particularly necessary. We wouldn't have to do packet buffering, for example, which alone is a very complex feature. It has the advantage of not needing to scan the network for TCP connections, which should improve the reliability of the transport. It may eliminate the need for proposal 0142 entirely, as there would never be two different apps connecting on two different transports; since IAP would always be the "primary" transport. This may still be necessary for Android however.

The only downside I can see is that a head unit could not **only** support TCP and not support IAP at all. However, due to the massive downsides of only supporting TCP (if the user's screen turns off their app goes away, media apps don't work, etc.), I think that's an acceptable requirement.

### Receiver Report
The `receiverReport` may encounter issues on iOS. Scheduled repeating timers do not always work well in the background (at least they did not when `heartbeat` was used). There may need to be a workaround (e.g. the head unit always sends a receiver report request with their latest and the proxy always responds with its latest, this way the head unit is in charge).

### Additional Notes
Because `SDLProtocol` and `SDLProxy` are public file in the iOS proxy, the addition of the `SDLProtocolSelector` and change of functionalities would likely constitute a major version change. It's been accepted that the protocol layer should be made private via a major version change, but I did want to note that.

The design allows the developer to choose which transports they can use, I believe they should not have this choice. It should be managed strictly and only by us. Even if it is decided that the developer should have this choice, it should happen in the LifecycleConfiguration, not via the API described in this proposal.

The modification to HMI_API.xml is breaking and requires a major version change of the HMI_API and Core.

**Comment by shoamano83** [comment]
Thanks @joeljfischer for your review comments.

> Alternate Transport Flow

This is a very good point. The proposal is created in mind that we will pursue the use-case of utilizing just Wi-Fi transport, especially on Android phones. We are aware that on iOS, iAP transport is almost mandatory for production apps. Perhaps @robinmk can tell the background a little more.
Also, we want apps to be transport agnostic on Android. Ideally, we don't want to say like "this transport is mandatory and others are optional".

> It may eliminate the need for proposal 0142 entirely, as there would never be two different apps connecting on two different transports; since IAP would always be the "primary" transport. This may still be necessary for Android however.

I'll reply to this comment in SDL 0142.

> The receiverReport may encounter issues on iOS. Scheduled repeating timers do not always work well in the background (at least they did not when heartbeat was used).

Thanks for the info. The idea is to send `Receiver Report` frames periodically from both Core and Proxy. I don't know the optimal value for an interval of the frames, but I think 500 or 1000 msecs will work. So iOS Proxy will be woken up periodically when it receives frames from Core, and I think it will have chances to send frames while it is awake.
By the way, do you know the interval and timeout value used for `HeartBeat`?

> Because SDLProtocol and SDLProxy are public file in the iOS proxy, the addition of the SDLProtocolSelector and change of functionalities would likely constitute a major version change.

OK thanks, I wasn't aware that SDLProtocol is public. I'm looking at iOS Proxy's code now and I would like to keep consistency of SDLProxy's public interface as much as possible.

> The design allows the developer to choose which transports they can use, I believe they should not have this choice. It should be managed strictly and only by us.

Right now it is up to each app developer to decide which transport is used. I thought I should not break this basic design, and that was why I proposed an API like this. I'm completely fine to switch to an idea in which Proxy selects the transport instead of app. I just want to know if everyone is OK with it (since that seems to be a major change).

> Even if it is decided that the developer should have this choice, it should happen in the LifecycleConfiguration, not via the API described in this proposal.

Regarding the LifecycleConfiguration, I think you are right. I was looking at the methods in ProxyManager.m, which are example codes.

> The modification to HMI_API.xml is breaking and requires a major version change of the HMI_API and Core.

Yes. Actually this proposal was developed prior to the discussion in SDL 0138. I think we may need to come up with another HMI_API idea which has less impact.

**Comment by robinmk** [comment]
> Alternate Transport Flow

From what I understand, this design sacrifices the ability for apps from multiple Android devices connected over Wi-Fi from registering since there is a dependency on BT. However, the design does indeed greatly reduce the complexity of the feature.
We will make an internal assessment on sacrifice in the use cases vs reduction in complexity and provide our feedback.

**Comment by joeljfischer** [comment]
> By the way, do you know the interval and timeout value used for HeartBeat?

I believe it was 5 seconds.

> OK thanks, I wasn't aware that SDLProtocol is public. I'm looking at iOS Proxy's code now and I would like to keep consistency of SDLProxy's public interface as much as possible.

It's been in the works for quite a while now to hide the SDLProxy, Protocol, and Transport layers by making them private. That should happen this year anyway. I'm just making everyone aware that by the time this was implemented, this would have to be done or it would be a breaking change.

> Right now it is up to each app developer to decide which transport is used. I thought I should not break this basic design, and that was why I proposed an API like this. I'm completely fine to switch to an idea in which Proxy selects the transport instead of app. I just want to know if everyone is OK with it (since that seems to be a major change).

*Technically*, yes. The developer can choose, however, they only have a choice between a debug and a production transport. I think this should stay, and a developer shouldn't have to choose between production transports. To put that decision on the developer is a recipe for confusion.

> Yes. Actually this proposal was developed prior to the discussion in SDL 0138. I think we may need to come up with another HMI_API idea which has less impact.

Thanks, this is important to note.

**Comment by robinmk** [comment]
> Alternate Transport Flow

We agree that the simplicity offered by the design mentioned by Joel outweighs the limits imposed on the certain use-cases. We will work on creating proposals which explains the design more in depth.

**Comment by theresalech** [comment]
The Steering Committee agreed that this proposal should be returned for revisions, to adjust the approach so that there is a primary transport identified before connecting over a WiFi transport to start specific services on, instead of connecting to all available transports and having all sessions run on all transports.  After discussing the differences in scope, effort, feasibility, and usability to allow WiFi as a primary transport as opposed to secondary, it was determined that we will start with the more simplified design outlined in joeljfischer's comments (where BT or USB serve as a primary transport), and can enter a separate proposal in the future to enable WiFi as a primary transport.

**Comment by theresalech** [comment]
The author has updated this proposal per the agreed upon revisions, and the revised proposal has now moved into review.  The Steering Committee will vote on this revised proposal during the 2018-03-13 meeting.

**Comment by joeygrover** [comment]
>After Secondary Transport is established, Proxy sends a Start Service frame for RPC service on it. Note that Proxy and Core don't run RPC service on Secondary Transport. 

I don't agree with this. It seems like a hack rather than a proper solution. I believe it is fine to create a new control message to handle this situation if needed, but Core should be able to handle a StartService message of any service as long as the session id matches the previously established session ID of the RPC service.  

>Also, if possible Proxy should retry setting up Secondary Transport on a regular interval.

The proxy should only retry if the secondary transport becomes available again via a notification from Core.

>To set up a TCP transport, Proxy needs to know the IP address and TCP port number on which Core is listening. A new Control Frame is proposed to convey the information. It is called Transport Config Update frame and is sent by Core to Proxy on Primary Transport. It should not be sent prior to Version Negotiation.

I believe this could be named better, something like `TransportEventUpdate` that held the status of the available transports. This would include any relevant information about the transport.

Also, the payloads in the control messages should be flat. This is by design as to keep the packets brief and not try to replicate RPC functionality. 

> `secondaryTransport`  &`servicesMap`
I believe these fields become redundant. Wouldn't it make more sense to just have a map of services to array of transports prefered. Again, I have to encourage the flat design of the payload
```
{
    "CONTROL_SERVICE_TRANSPORTS": [TCP_WIFI, SPP_BLUETOOTH],
    "VIDEO_SERVICE_TRANSPORTS": [TCP_WIFI]
}
```

>Extension of Android Proxy

The Android proxy implementation is likely to be a bit difficult due to the way in which transports and sessions interact. We might need a separate proposal to refactor a lot of this.

**Comment by joeljfischer** [comment]
> After Secondary Transport is established, Proxy sends a Start Service frame for RPC service on it. This procedure is just for Core to recognize that Proxy initiates Secondary Transport, because Core cannot know which transport belongs to which app(s) at transport layer's level.

For what reason can this not be a start service of a transport the proxy actually wishes to start (using the same session id)? If that will not work, I would recommend a different protocol message.

> When Secondary Transport becomes unavailable, Core and Proxy should abort any services that are running on Secondary Transport. Also, if possible Proxy should retry setting up Secondary Transport on a regular interval.

I think this should be more deterministic. If Proxy knows the secondary transport name, it would happen when it re-enters the foreground, for example.

> iOS Proxy implementation should include:

I think the iOS Proxy implementation will have to include taking the current TCP transport from a debug to a production state.

**Comment by shoamano83** [comment]
Hi @joeygrover, @joeljfischer, thanks for your comments. First, let me reply to the comments regarding Start Service procedure on Secondary Transport.

> I don't agree with this. It seems like a hack rather than a proper solution. I believe it is fine to create a new control message to handle this situation if needed, but Core should be able to handle a StartService message of any service as long as the session id matches the previously established session ID of the RPC service.

Yes, even if we don't run Start Service / Start Service ACK sequence of RPC service on Secondary Transport, Core can handle Start Service message of any services using the previous Session ID.

The reason that I would like to have Start Service / Start Service ACK sequence of RPC service on Secondary Transport is that Core has to notify HMI of Secondary Transport's connection through `BC.UpdateAppList`. The event should be notified before a service (like video service) is started on Secondary Transport. Unfortunately, Core cannot recognize which transport is used by which app until a service is started on that transport. That is why I would like to start a service after Secondary Transport is established, and I thought RPC service is suitable.

We may be able to add a control frame for the event, but from Core's implementation point of view handling the control frame will likely to be similar to handling Start Service frame. That didn't sound like a good idea.

> For what reason can this not be a start service of a transport the proxy actually wishes to start (using the same session id)? If that will not work, I would recommend a different protocol message.

Core's implementation isolates transports and services. They are associated when a service is started on a transport (using Start Service / Start Service ACK). So, my understanding is that Core requires a service is started on the Secondary Transport to recognize that an app has created a Secondary Transport.

Sorry I didn't get your comment completely, but if we just run Start Service / Start Service ACK again on the Primary Transport, Core will not recognize Secondary Transport.

I agree that the idea sounds odd since we are not actually running a RPC service on Secondary Transport. An alternative can be to add a "Null" or "Transport Management" service in the protocol spec which actually does nothing. Proxy starts the service once Secondary Transport is established.

**Comment by shoamano83** [comment]
> The proxy should only retry if the secondary transport becomes available again via a notification from Core.

I meant that if Core didn't notify transport unavailability and Secondary Transport is disconnected, then Proxy should retry connection. This happens, for example, when air condition becomes bad and Wi-Fi link gets unstable. When Core notifies transport unavailability (for example when the user turned off Wi-Fi feature on the head unit) then Proxy will not retry connection. Does it make sense?

> I believe this could be named better, something like `TransportEventUpdate` that held the status of the available transports.

I don't mind renaming the frame. Let us use it unless we get an objection or better naming from members.

> Also, the payloads in the control messages should be flat. This is by design as to keep the packets brief and not try to replicate RPC functionality.

OK, so you mean that the bson object should be like this?
```
{
  "tcpIpAddress": "192.168.1.1",
  "tcpPort": 12345
  /* if we add information of another transport in future, then it will be added here */
}
```

> I believe these fields become redundant. Wouldn't it make more sense to just have a map of services to array of transports prefered.

The intention of `secondaryTransport` field is to explicitly specify the transport to use as Secondary. Having just a map seemed OK at first, but I realized that it would cause an issue if the .ini file is mistakenly configured like this:
```
{
  "CONTROL_SERVICE_TRANSPORTS": ["TCP_WIFI", "AOA_USB", "SPP_BLUETOOTH"],
  "VIDEO_SERVICE_TRANSPORTS": ["TCP_WIFI", "AOA_USB"],
  "AUDIO_SERVICE_TRANSPORTS": ["AOA_USB", "TCP_WIFI"]
}
```

In this example video service prefers Wi-Fi but audio prefers USB, so Proxy cannot tell which transport it should start. I wanted to make the fields less ambiguous to avoid such confusion. (By the way, is Control Service still used? I thought it was only for HeartBeat frames.)

I am fine to make the bson object flat, so how about this?
```
{
  "secondaryTransport": "TCP_WIFI",  // this field will be omitted if Core decides that Secondary Transport is unnecessary
  "videoServiceTransports": [2],     // 1 for Primary, 2 for Secondary. We don't specify transport name like "IAP_USB" since Proxy has to know whether the iAP transport is USB or Bluetooth.
  "audioServiceTransports": [2]
}
```

**Comment by shoamano83** [comment]
> I think this should be more deterministic. If Proxy knows the secondary transport name, it would happen when it re-enters the foreground, for example.

Again, the sentence was written in mind that Wi-Fi link can get unstable and TCP transport may be disconnected. In such case, I thought that TCP transport should be restored without any events (like notification from Core or app's state change to foreground.)

By the way, I agree that iOS Proxy should stop Wi-Fi Secondary Transport when the app goes to background, then should restart the transport when the app becomes foreground. I think this has been covered by "Extension of iOS Proxy" section.

> I think the iOS Proxy implementation will have to include taking the current TCP transport from a debug to a production state.

OK, could you suggest what tasks we will likely need for this? (For example, more code reviews on existing SDLTCPTransport?)

Thanks.

**Comment by AndrewRMitchell** [comment]
Re potential downsides, specifically Bluetooth/Wi-Fi interference, another wireless car interface basically requires that either the Wi-Fi transport be done at 5 GHz or the Bluetooth connection be turned off once Wi-Fi is established (and recommends doing this at 5 GHz anyway). I understand why we're specifying things the way we are, but I am a bit concerned we simply leave the 2.4 GHz band as a potential downside. In today's world (and given the amount of 2.4 GHz interference), why not just require 5 GHz as part of this proposal.

**Comment by shoamano83** [comment]
Hi @AndrewRMitchell , thanks for your comment. I don't know if we can make 5GHz mandatory as requirements for 5GHz band by law differ between countries. Maybe we can "recommend" using 5GHz Wi-Fi in the proposal document if an OEM wants to implement and enable this feature.

**Comment by shoamano83** [comment]
> I agree that the idea sounds odd since we are not actually running a RPC service on Secondary Transport. An alternative can be to define a "Null" or "Transport Management" service in the protocol spec which actually does nothing. Proxy starts the service once Secondary Transport is established.

During today's steering committee call Alex suggested that Proxy can start video service after Secondary Transport even if the app's HMI level is not FULL. The Start Service frame will tell Core that Secondary Transport is established, while Core can reply with Start Service NACK if the app's HMI level is inappropriate. I think this is another approach that should work, so I would like feedback on this idea from other members as well.

My concern is that it will impact Proxy:
- iOS Proxy checks all necessary condition (HMIlevel and app's foreground/background status) in the lifecycle manager before starting video/audio service. The idea will impact existing implementation.
- Android Proxy starts video service when the app requests starting video service through API call. (My understanding is not up-to-date so please correct me if I'm wrong.) So it may not make sense if Proxy automatically starts video service without app's request; even if the service is accepted and video streaming starts, the app may not want to send out any video frames.

**Comment by joeygrover** [comment]
### Initial secondary transport connection

After the call, I believe I understand the use case for notifying Core/HMI that an app has "registered" on the new transport. While sending a `StartService` packet introduces less change to the protocol, I think it still feels a little bit of a stretch. I still believe a new control frame should be introduced, maybe something like `RegisterSecondaryTransport` and corresponding ACK message. Then in the proxies when they receive the ACK they will know it is safe to start services on that transport. 

### TransportEventUpdate

Still open to a better name. Yes, I prefer the flat structure, but if others do not agree then comments are welcome.

### Retry
I believe we can flush out some of this during implementation, but overall this is what I think would be appropriate

1. Secondary transport connection becomes unstable and disconnects.
2. Core notifies the proxy via `TransportEventUpdate` that the connection was lost, but proxy should retry
3. If Core says not to retry, skip to **6**
4. Proxy attempts to reconnect x amount of times
5. If successful, continue normal flow
6. If not, inform user using Alert or similar on primary channel to try again later. 

Possibly when the user attempts to bring the app into HMI_FULL, proxy could try to connect to the secondary transport again. 

### `StartService` Secondary transport tags 

Does it actually make more sense to include this information into the `TransportEventUpdate` control frame? It seems like this information might get updated based on the availability of different transports and therefore best included into the frame that is sent based on that information. 

For the payload, I believe what you commented will work. Would we want to use 1 = primary, 2 = secondary, 3 = either? Also, should we consider secondary transports could be an array?

**Comment by shoamano83** [comment]
Hi @joeygrover, Thanks for your feedback.

> ### Initial secondary transport connection

OK, while I still prefer employing Start Service / Start Service ACK pair (possibly with a new service like "Transport Management" service), I'm OK with utilizing a new control frame if the majority of the members prefer that approach.

> ### TransportEventUpdate

Understood. Actually I started implementing this in iOS Proxy and found that multi-level parameter looks a little more confusing than I expected. If we are not adding many parameters in the frame, a flat design may be preferable.

> ### Retry

I agree that we can work the details out during implementation. For now, let me just leave a few comments.

If I understand correctly, the key points of the sequence proposed by Joey are:
1. Core initiates the retry sequence by notifying Proxy
2. User will be notified if retrying connection didn't work

For point #1, will we have a case that Core doesn't want to retry? (Actually I didn't come up with any.) If Core always initiates retry after disconnection, then I think we can let Proxy initiate retry by itself. This approach may look simpler.
For point #2, I agree that some people like to get such error notification. And such notification may be useful in the case AOA is used for Secondary Transport (i.e. the user can try reconnecting the cable). However, in the case of Wi-Fi, there is little that the user can do to improve the air condition issue.

Let me also note that when Core detects disconnection of Secondary Transport, HMI will be notified of the event. So if we need an error message, I think HMI can include implementation to wait for XX seconds after disconnection and provide a gentle warning to check cable (if previous Secondary Transport connection was USB) or to check air condition (if it was Wi-Fi). This approach will not introduce a notification between Proxy and Core.

> ### `StartService` Secondary transport tags

> Does it actually make more sense to include this information into the `TransportEventUpdate` control frame? 

That should also work, as long as `videoServiceTransports` and `audioServiceTransports` are kept in Start Service ACK frame. My concern is that it will introduce complexity in Core code, as Core needs to detect updates on each transport's availability and reflect it to `Transport Event Update` frame. So I prefer making `secondaryTransport` param "static" in Start Service ACK, i.e. it will not be updated based on transports' availabilities. If we need to switch Secondary Transport for example between Wi-Fi and AOA, I think we can make `secondaryTransport` array instead. (See my comment below.)

> Would we want to use 1 = primary, 2 = secondary, 3 = either?

That is fine as long as "3" implies that Secondary Transport is always preferred over Primary. My preference is to keep the values as enum, but I don't have strong opinion on it.

> Also, should we consider secondary transports could be an array?

I think this idea is to suggest a list of candidate transports to Proxy, and let it switch Secondary Transport between them. As I put in the "Alternatives considered" section, right now I don't think such advanced feature is necessary. On the other hand, if we are concerned about updating protocol spec again in future, I am OK to make it an array (also renaming it to `secondaryTransports`) and put a limitation that we only support one element in the the array for now.

Thanks!

**Comment by joeljfischer** [comment]
> OK, could you suggest what tasks we will likely need for this? (For example, more code reviews on existing SDLTCPTransport?)

I think this would likely entail an entire rewrite of the TCP Transport using https://developer.apple.com/documentation/foundation/nsurlsessionstreamtask

**Comment by theresalech** [comment]
The Steering Committee has voted to accept this proposal with revisions. The revisions are as follows:

- Introduce a new Control Frame, `RegisterSecondaryTransport`, which will be used once a Secondary Transport is connected to notify Core that the Proxy has established a connection and tell Core which session is registering.
- Rename `Transport Config Update` to `TransportEventUpdate`. The original parameters will be refactored to a flat structure.
- StartServiceACK for the RPC service will contain a Secondary Transport param that will contain an array of potential secondary transports instead of a single value. The service to transport priority parameters will become flat instead of document style; each service param will contain an ordered array in terms of priority of potential transports.
- Proxy connection retry strategy will include the new `TransportEventUpdate` Control Frame. When a connection is interrupted/ended, a `TransportEventUpdate` control frame will be sent to the proxy on the primary transport. If the proxy is to retry (in case of WiFi), it will include the IP address and Port. If the proxy is not to retry, the frame will contain either empty strings for IP address and port or not include the parameters at all.
- Will recommend, but not require, using 5GHz
- Proxy implementations will be discussed outside of this proposal.

**Comment by theresalech** [comment]
@shoamano83 or @robinmk please advise when a new PR has been entered to update the proposal to reflect the agreed upon revisions. I'll then merge the PR so the proposal is up to date, and enter issues in the respective repositories for implementation. Thanks!

**Comment by theresalech** [comment]
PR with revisions has been merged and issues have been entered:
[Core](https://github.com/smartdevicelink/sdl_core/issues/2065)
[RPC](https://github.com/smartdevicelink/rpc_spec/issues/68)
[iOS](https://github.com/smartdevicelink/sdl_ios/issues/900)
[Android](https://github.com/smartdevicelink/sdl_android/issues/714)
